[browser][wasm] Add support for marshaling Enum through the bindings

kjpou1 opened the pull request:
Initial support for marshaling Enum types through the bindings.

Closes https://github.com/dotnet/runtime/issues/40188

Javascript does not have support for Enum data types and instead ordinarily uses strings to represent property values.  Although, this is not always true as they can be [typed as well](https://javascript.info/dom-attributes-and-properties#dom-properties-are-typed) most property values can be expressed as strings.

An example is the [Color property](https://www.w3schools.com/jsref/prop_style_color.asp).

The marshaling code allows one to set a binding contract through the `Export` Attribute to help facilitate the marshaled values.

```
    public enum ConvertEnum
    {
        Default,
        ToLower,
        ToUpper,
        Numeric
    }
```

example:

```
        public enum StyleColor : uint
        {
           RedNoExport,
            [Export(EnumValue = ConvertEnum.Default)]
            Red = 0xff0000ff,
            [Export(EnumValue = ConvertEnum.Numeric)]
            GreenNumeric = 0x00ff00ff,
            [Export("red", EnumValue = ConvertEnum.ToUpper)]
            RedUpperCase,
            [Export("RED", EnumValue = ConvertEnum.ToLower)]
            RedLowerCase,
            [Export("#ff0000")]
            RedHex
        }
```

- No `Export` attribute.  `RedNoExport ` is marshaled as the string `RedNoExport `
- ConvertEnum.Default uses the Enum value itself.  `Red` is marshaled as the string `Red`
- ConvertEnum.ToLower will convert the enum value to lower case before marshaling the value. `Red` is marshaled as the string `red`.
- ConvertEnum.ToUpper will convert the enum value to upper case before marshaling the value. `Red` is marshaled as the string `RED`.
- ConvertEnum.Numeric will marshal the numeric value instead of the string value.  'GreenNumeric' will be marshaled as 16711935

One can also marshal a set value as well by using the `ContractName` parameter of the `Export` attribute.

```
            [Export("#ff0000")]
            RedHex
```

Will be marshaled as `#ff0000`.

The `Export` attribute also facilitates the marshaling of these values from JS to managed code.

**** Note **** 

Restriction on Enum type of long and ulong exist as JavaScript does not at this time have support for native int64 values.

kjpou1 committed:
[browser][wasm] Add support for marshaling Enum through the bindings

kjpou1 committed:
Fix invalid cast on ConvertEnum.Numeric case

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-enum-bindings

kjpou1 committed:
Fix error message when enum's are defined as long or ulong

kjpou1 committed:
Fix enum marshaling by numeric value contract

kjpou1 committed:
Fix parse exception and add more tests.

kjpou1 committed:
Add test entry for Enum without an Export attribute.

marek-safar commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Export.cs:
```diff
@@ -0,0 +1,43 @@
+
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Property | AttributeTargets.Method | AttributeTargets.Field,
+            AllowMultiple = true, Inherited = false)]
+    public class ExportAttribute : Attribute
```
The name is odd if we are using it for enums only

kjpou1 replied:
Ok will change it to EnumExport.

kjpou1 replied:
Addressed in recent commits

marek-safar commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,130 @@ private static DateTime CreateDateTime(double ticks)
             return unixTime.DateTime;
         }
 
+        private static object? ObjectToEnum(IntPtr methodHandle, int parm, object obj)
```
Cannot we pass parameter type directly?

kjpou1 replied:
Not sure will need to check but do not think we have it.

marek-safar commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,130 @@ private static DateTime CreateDateTime(double ticks)
+
+        private static Enum EnumFromExportContract(Type enumType, object value)
+        {
+            if (!enumType.IsEnum)
```
This should be just Assert

kjpou1 replied:
Addressed in recent commits.

marek-safar commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,130 @@ private static DateTime CreateDateTime(double ticks)
+            }
+        }
+
+        private static object EnumToExportContract(Enum value)
```
Why to return an object type if we always need a string?

kjpou1 replied:
It can also return a primitive value as well.

marek-safar replied:
Sure but why should it return as object if we immediate convert it to string?

marek-safar commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Export.cs:
```diff
@@ -0,0 +1,43 @@
+        public string? ContractName { get; }
+
+        public Type? ContractType { get; }
+        public ConvertEnum EnumValue { get; set; }
```
Is this really needed? It does not really add a lot of value and adds unnecessary dependency on reflection and metadata

kjpou1 replied:
Addressed in recent commit.  Removed all non essential code elements.  Let me know if this is sufficient.

marek-safar commented on src/mono/wasm/runtime/corebindings.c:
```diff
@@ -165,11 +165,13 @@ mono_wasm_unbox_enum (MonoObject *obj)
-	// case MONO_TYPE_I8:
-	// case MONO_TYPE_U8:
+	// WASM doesn't support returning longs to JS i.e. JavaScript does not have a native 64-bit int
+	case MONO_TYPE_I8:
+	case MONO_TYPE_U8:
+		fprintf (stderr, "Invalid type %d to core mono_unbox_enum due to JavaScript not having a native 64-bit int.\n", mono_type_get_type(mono_type_get_underlying_type (type)));
+		return 0;
 	default:
-		printf ("Invalid type %d to mono_unbox_enum\n", mono_type_get_type(mono_type_get_underlying_type (type)));
+		fprintf (stderr, "Invalid type %d to mono_unbox_enum\n", mono_type_get_type(mono_type_get_underlying_type (type)));
```
Please move these into separate PR

kjpou1 replied:
Added PR: https://github.com/dotnet/runtime/pull/41437

marek-safar commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Export.cs:
```diff
@@ -0,0 +1,43 @@
+        {
+        }
+
+        public ExportAttribute(Type contractType) : this(null, contractType)
```
What is the scenario for this overload?

kjpou1 replied:
This is removed now.

marek-safar commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/EnumMarshalTests.cs:
```diff
@@ -0,0 +1,94 @@
+
+namespace System.Runtime.InteropServices.JavaScript.Tests
+{
+    public static class EnumMarshalTests
```
Please add tests where
	- multiple Export attributes exist
	- Export with the same value is used for multiple enum values
	- values don't match to any of enum values

marek-safar commented:
Is this needed for 5.0 ?

kjpou1 commented:
I would say no it is not needed for 5.0.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-enum-bindings

kjpou1 committed:
Address review comments The name is odd if we are using it for enums only

- Rename Export to EnumExport

kjpou1 committed:
Address review to use Assert instead of throwing an exception

kjpou1 committed:
Remove non essential code elements from attribute.

kjpou1 committed:
This code is moved to a separate PR: https://github.com/dotnet/runtime/pull/41437

marek-safar commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/EnumExport.cs:
```diff
@@ -0,0 +1,32 @@
+        {
+        }
+
+        public EnumExportAttribute(string? contractName)
```
Should be just `string`

kjpou1 replied:
Addressed in recent commits

marek-safar commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/EnumExport.cs:
```diff
@@ -0,0 +1,32 @@
+            AllowMultiple = true, Inherited = false)]
+    public class EnumExportAttribute : Attribute
+    {
+        public EnumExportAttribute() : this(null)
```
No need to call this(null)

kjpou1 replied:
Addressed in recent commits

marek-safar commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/EnumExport.cs:
```diff
@@ -0,0 +1,32 @@
+        }
+
+        public string? ContractName { get; }
+        public ConvertEnum EnumValue { get; set; }
```
I'm still not convinced this extra logic dimension is useful

/cc @lewing

kjpou1 replied:
This has all been simplified now.  

Simplify the export interface. 
- remove extra logic dimension
- remove unnecessary dependency on reflection and metadata
- default is to now return numeric value if no attribute is defined.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-enum-bindings

radical commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/EnumMarshalTests.cs:
```diff
@@ -0,0 +1,94 @@
+            Red = 0xff0000ff,
+            [EnumExport(EnumValue = ConvertEnum.Numeric)]
+            GreenNumeric = 0x00ff00ff,
+            [EnumExport("red", EnumValue = ConvertEnum.ToUpper)]
```
Just wondering, is this a useful use case? Maybe we should have just two modes:
1. default - returns either `ContractName`, or the field name as a string
2. number - returns the numeric value of the enum

kjpou1 replied:
Simplified no longer valid

radical commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/EnumMarshalTests.cs:
```diff
@@ -0,0 +1,94 @@
+        }
+
+        [Fact]
+        public static void MarshalRequestEnums()
```
nit: Should this be renamed to `StringFromJSToManagedEnum()`? or `JSStringToManaged`, so this sounds like `JSString` is some type being used here.

kjpou1 replied:
Not sure I understand why this is needed or a nit.  Could you explain this more as to why it is needed.

radical replied:
Just as a more descriptive name for the test. `MarshalStringFromJSToManagedEnum`?

kjpou1 replied:
Gotcha.  Addressed in recent commits

radical commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/EnumMarshalTests.cs:
```diff
@@ -0,0 +1,94 @@
+        }
+
+        [Fact]
+        public static void MarshalRequestEnumProps()
```
And this one to `SetPropertiesOnJSObjectFromManaged`?

kjpou1 replied:
Not sure I understand why this is needed or a nit.  Could you explain this more as to why it is needed.

radical replied:
Just as a more descriptive name for the test.

kjpou1 replied:
Gotcha. Addressed in recent commits

radical commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/EnumMarshalTests.cs:
```diff
@@ -0,0 +1,94 @@
+            Assert.Equal(StyleColor.RedNoExport, HelperMarshal.styleColorEnums[5]);
+        }
+    }
+}
```
Would it be useful to have tests for enums being returned from methods/functions?
How about a test for array of enums being marshaled?

kjpou1 replied:
Array of enums is not supported right now nor native arrays in general.  They all use a JavaScript object to marshal back and forth.  We can set up a PR for an enhancement to support native arrays but I have not found any examples that actually use an array of enums.  What could be done is create a JavaScript.Array object and fill it out with what is needed to be passed.

radical commented on src/mono/wasm/runtime/binding_support.js:
```diff
@@ -123,6 +123,8 @@ var BindingSupportLib = {
 			this.safehandle_get_handle = get_method ("SafeHandleGetHandle");
 			this.safehandle_release_by_handle = get_method ("SafeHandleReleaseByHandle");
 
+			this.object_to_enum = get_method ("ObjectToEnum");
```
Just trying to understand - I see `object_to_enum` being used in `js_to_mono_enum`, but it is being added here. How was it working earlier?

kjpou1 replied:
It was not working before.  Thus the issue and pr

kjpou1 committed:
Address review comments

radical reviewed (changes requested):
I'm wondering whether all the field lookups need to be cached. Or maybe special-case marshaling for array-of-enums?

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,127 @@ private static DateTime CreateDateTime(double ticks)
+
+            object? contractName = value;
+
+            if (attributes != null && attributes.Length > 0)
```
We can throw an error if there are multiple ones.  Is that preferable?

radical replied:
If we don't want multiple then we can remove `AllowMultiple` from the `AttributeUsage`.

kjpou1 replied:
Addressed in recent commits

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-enum-bindings

kjpou1 committed:
Rename source module as per review comments.

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System.Private.Runtime.InteropServices.JavaScript.csproj:
```diff
@@ -34,6 +34,7 @@
     <Compile Include="System\Runtime\InteropServices\JavaScript\HostObject.cs" />
     <Compile Include="System\Runtime\InteropServices\JavaScript\Function.cs" />
     <Compile Include="System\Runtime\InteropServices\JavaScript\AnyRef.cs" />
+    <Compile Include="System\Runtime\InteropServices\JavaScript\EnumExport.cs" />
```
Addressed in recent commits

kjpou1 committed:
Address review comments about AttributeUsage

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/EnumExport.cs:
```diff
@@ -0,0 +1,32 @@
+        Numeric
+    }
+
+    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Property | AttributeTargets.Method | AttributeTargets.Field,
```
Not `AttributeTargets.Enum` but  `AttributeTargets.Field` makes more sense.  Addressed in recent commits

kjpou1 committed:
remove the actual value `AllowMultiple = false`

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-enum-bindings

kjpou1 committed:
Simplify the export interface.

- remove extra logic dimension
- remove unnecessary dependency on reflection and metadata
- default is to now return numeric value if no attribute is defined.

kjpou1 committed:
Simplify  to string

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -379,6 +379,11 @@ void Complete()
         {
+            if (o is Enum oIsAnEnum)
+            {
+                return EnumToExportContract(oIsAnEnum)?.ToString() ?? string.Empty;
```
Done

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,127 @@ private static DateTime CreateDateTime(double ticks)
+                    contractName = contractName!.ToString()!.ToUpper();
+                    break;
+                case ConvertEnum.Numeric:
+                    contractName = Convert.ToDouble(Enum.Parse(value!.GetType(), contractName!.ToString() ?? string.Empty));
```
done

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/EnumExport.cs:
```diff
@@ -0,0 +1,32 @@
+        Default,
+        ToLower,
+        ToUpper,
+        Numeric
```
No longer valid

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,127 @@ private static DateTime CreateDateTime(double ticks)
+                    break;
+            }
+
+            return contractName ?? string.Empty;
```
no longer valid

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,127 @@ private static DateTime CreateDateTime(double ticks)
             return unixTime.DateTime;
         }
 
+        private static object? ObjectToEnum(IntPtr methodHandle, int parm, object obj)
```
???

radical replied:
Just a suggestion! I thought it might be a typo, so suggested `param` instead of `parm`.

kjpou1 replied:
Gotcha.  Addressed in recent commits

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -379,6 +379,11 @@ void Complete()
 
         private static string ObjectToString(object o)
         {
+            if (o is Enum oIsAnEnum)
```
???

radical replied:
Just a suggestion! `oIsAnEnum` just sounds like a bool to me, so I was suggesting changing the name.

kjpou1 replied:
Gotcha. Addressed in recent commits

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-enum-bindings

kjpou1 committed:
Address review comments about variable and parameter names

kjpou1 committed:
Address review comments about name of test method

radical commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/EnumMarshalTests.cs:
```diff
@@ -0,0 +1,95 @@
+            OnlyIfCached = -3636,
+        }
+
+        public enum StyleColor : uint
```
And tests for invalid values.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-enum-bindings

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,94 @@ private static DateTime CreateDateTime(double ticks)
+                    {
+                        if (Enum.TryParse(enumType, value!.ToString() ?? string.Empty, out object? enumAsNumeric))
+                        {
+                            if (enumAsNumeric is Enum valueAsEnum)
+                                return (Enum)valueAsEnum;
```
doing that results in:  Converting null literal or possible null value to non-nullable type.

radical replied:
Maybe `return (Enum)enumAsNumeric! ;` ?

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,94 @@ private static DateTime CreateDateTime(double ticks)
+
+                    if (contractName!.ToString() == value.ToString())
+                    {
+                        return (Enum)Enum.Parse(enumType, fi.Name);
```
This results in Converting null literal or possible null value to non-nullable type.

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,94 @@ private static DateTime CreateDateTime(double ticks)
+
+            if (attributes?.Length > 0)
+            {
+                contractName = attributes[0].ContractName;
```
not valid as gives Possible null reference return.

radical replied:
If you decide that `EnumExportAttribute.ContractName` cannot be null, then this can be done.

kjpou1 committed:
Address review comments

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,94 @@ private static DateTime CreateDateTime(double ticks)
+                    }
+                    else
+                    {
+                        if (Enum.TryParse(enumType, value!.ToString() ?? string.Empty, out object? enumAsNumeric))
```
Addressed in recent commits

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,94 @@ private static DateTime CreateDateTime(double ticks)
+
+                    object? contractName = null;
+
+                    if (attributes?.Length > 0)
```
Addressed in recent commits

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,94 @@ private static DateTime CreateDateTime(double ticks)
+            }
+            else
+            {
+                contractName = Convert.ToDouble(Enum.Parse(value!.GetType(), contractName!.ToString() ?? string.Empty)).ToString();
```
Addressed in recent commits

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -379,6 +379,11 @@ void Complete()
         {
+            if (o is Enum oAsAnEnum)
+            {
+                return EnumToExportContract(oAsAnEnum)?.ToString() ?? string.Empty;
```
Addressed in recent commits

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,94 @@ private static DateTime CreateDateTime(double ticks)
+                    }
+                    else
+                    {
+                        if (Enum.TryParse(enumType, value!.ToString() ?? string.Empty, out object? enumAsNumeric))
```
Addressed in recent commits

kjpou1 committed:
Add support and test for `[ExportEnum] MyEportedEnumValue`

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/EnumExportAttribute.cs:
```diff
@@ -0,0 +1,21 @@
+    [AttributeUsage(AttributeTargets.Field, Inherited = false)]
+    public class EnumExportAttribute : Attribute
+    {
+        public EnumExportAttribute()
```
It is now.  Added support in recent commits.

kjpou1 committed:
address review comments

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,94 @@ private static DateTime CreateDateTime(double ticks)
             return unixTime.DateTime;
         }
 
+        private static object? ObjectToEnum(IntPtr methodHandle, int paramIdx, object obj)
```
Addressed in recent commits

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,94 @@ private static DateTime CreateDateTime(double ticks)
+
+        private static Enum EnumFromExportContract(Type enumType, object value)
+        {
+            System.Diagnostics.Debug.Assert(enumType.IsEnum, $"Type provided for parameter `{nameof(enumType)}` must be an Enum. Instead the following type was found: {enumType}.");
```
See comment https://github.com/dotnet/runtime/pull/41081#discussion_r477223594

kjpou1 committed:
Address review comments about index out of range.

kjpou1 commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -401,6 +406,127 @@ private static DateTime CreateDateTime(double ticks)
+            MethodBase? mb = MethodBase.GetMethodFromHandle(tmp.handle);
+            if (mb == null)
+                return null;
+            Type? parmType = mb.GetParameters()[parm]?.ParameterType;
```
Addressed in recent commits

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-enum-bindings

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-enum-bindings

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-enum-bindings

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-enum-bindings

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-enum-bindings

# Conflicts:
#	src/libraries/System.Private.Runtime.InteropServices.JavaScript/tests/System/Runtime/InteropServices/JavaScript/HelperMarshal.cs

kjpou1 commented:
Will close this for now as will need more analysis due to recent revamping of the bindings.

kg commented:
The stuff I'm working on now should resolve the rest of the Enum stuff, and the binding revamp added basic enum support
